src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketRequestExecutor.java

```diff
@@ -0,0 +1,145 @@
+     *
+     * @param url        url to connect to
+     * @param returnType type reference used when getting generified objects (such as pages)
+     * @param <T>        type to return
+     * @return a deserialized object of type T
+     * @see #makeGetRequest(HttpUrl, Class)
+     */
+    public <T> BitbucketResponse<T> makeGetRequest(HttpUrl url, TypeReference<T> returnType) {
+        return makeGetRequest(url, in -> objectMapper.readValue(in, returnType));
+    }
+
+    /**
+     * Make a GET request to the url given. This method will add authentication headers as needed.
+     * <em>Note!</em> this method <em>cannot</em> be used to retrieve entities that makes use of
+     * generics (such as {@link BitbucketPage}) for that use {@link #makeGetRequest(HttpUrl,
+     * TypeReference)} instead.
+     *
+     * @param url        url to connect to
+     * @param returnType class of the desired return type. Do note that if the type is generified
+     *                   this method will not work
+     * @param <T>        type to return
+     * @return a deserialized object of type T
+     * @see #makeGetRequest(HttpUrl, TypeReference)
+     */
+    public <T> BitbucketResponse<T> makeGetRequest(HttpUrl url, Class<T> returnType) {
+        return makeGetRequest(url, in -> objectMapper.readValue(in, returnType));
+    }
+
+    /**
+     * Makes a POST request to the given URL with given request payload.
+     *
+     * @param url,            the URL to make the request to
+     * @param requestPayload, JSON payload which will be marshalled to send it with POST.
+     * @param returnType,     Class of expected return type
+     * @param <T>,            Type of Request payload.
+     * @param <R>,            Return type
+     * @return
+     */
+    public <T, R> BitbucketResponse<R> makePostRequest(HttpUrl url, T requestPayload, Class<R> returnType) {
+        ObjectReader<R> reader = in -> objectMapper.readValue(in, returnType);
+        return httpRequestExecutor.executePost(url, credentials, marshall(requestPayload), response ->
+                new BitbucketResponse<>(response.headers().toMultimap(), unmarshall(reader, response.body())));
+    }
+
+    private void ensureNonEmptyBode(Response response) {
+        if (response.body() == null) {
+            log.debug("Bitbucket - No content in response");
+            throw new NoContentException(
+                    "Remote side did not send a response body", response.code());
+        }
+    }
+
+    private <T> BitbucketResponse<T> makeGetRequest(HttpUrl url, ObjectReader<T> reader) {
+        return httpRequestExecutor.executeGet(url, credentials,
+                response -> {
+                    ensureNonEmptyBode(response);
+                    return new BitbucketResponse<>(
+                            response.headers().toMultimap(), unmarshall(reader, response.body()));
+                });
+    }
+
+    private <T> String marshall(T requestPayload) {
```

Comment:
neither this, nor the `unmarshall` below handles null, either we need to do nullchecking above _in all places_ or we just move it down here (which I think would be neater), either way we need to handle it.

Reply:
This is still an issue, it will still throw a NPE from a few locations, I maintain moving the null check here and throwing a `NoContentException` from here (and `unmarshall`) is better as it is an expected exception.